GSA-TTS/FAC
# UEI Validation Waivers

# UEI Validation Waivers

Issue: Closes https://github.com/GSA-TTS/FAC/issues/4027

## Changes:
Adjusts UEI validation to account for and accept "waived" UEIs. Also, a very basic admin panel for UEI validation waivers.

Includes one migration to make the UeiValidationWaiver "uei" field non-unique, for cases where a UEI needs to be waived again after the first expires.

Does not implement UEI waiver expiration dates yet. A comment is left in the UEI validation to accommodate them.

## How to test:
1. Switch to this branch.
2. Run normally, but make sure migrations apply.
3. Verify that the UEI validation waiver table appears in the DB and in the admin panel
4. Run a UEI that doesn't exist or is totally invalid (unregistered, inactive, missing data, whatever). Verify that it doesn't pass UEI validation.
    - I use `SUPERC00LUE1`, with numbers instead of letters for "O" and "I". 
6. Add a UEI validation waiver for that nonexistent UEI
    - Remember that the letters "O" and "I" are invalid, and will continue to be even if waived.
    - It still has to be 12 letters long.
7. Verify that this waived UEI now passes. 

If you want, take the submission all the way to dissemination, and verify that it still passes. We only do a SAM check on audit creation, so as long as the UEI passes the schema checks, we're good. 

## Recording:

https://github.com/user-attachments/assets/0df5ef5f-4547-4a53-a1f1-ef8a25266499

## PR checklist: submitters


## PR checklist: reviewers


The larger the PR, the stricter we should be about these points.
